mysqld_exporter.go

```diff
@@ -1,311 +1,347 @@
 			Namespace: namespace,
-			Name:      "exporter_scrapes_total",
-			Help:      "Current total mysqld scrapes.",
+			Subsystem: exporter,
+			Name:      "scrapes_total",
+			Help:      "Total number of times MySQL was scraped for metrics.",
 		}),
 		error: prometheus.NewGauge(prometheus.GaugeOpts{
 			Namespace: namespace,
-			Name:      "exporter_last_scrape_error",
-			Help:      "The last scrape error status.",
+			Subsystem: exporter,
+			Name:      "last_scrape_error",
+			Help:      "Whether the last scrape of metrics from MySQL resulted in an error (1 for error, 0 for success).",
 		}),
-		metrics: map[string]prometheus.Gauge{},
-		commands: prometheus.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: globalStatus,
-			Name:      "commands_total",
-			Help:      "Number of executed mysql commands.",
-		}, []string{"command"}),
-		connectionErrors: prometheus.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: globalStatus,
-			Name:      "connection_errors_total",
-			Help:      "Number of mysql connection errors.",
-		}, []string{"error"}),
-		innodbRows: prometheus.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: globalStatus,
-			Name:      "innodb_rows_total",
-			Help:      "Mysql Innodb row operations.",
-		}, []string{"operation"}),
-		performanceSchema: prometheus.NewCounterVec(prometheus.CounterOpts{
-			Namespace: namespace,
-			Subsystem: globalStatus,
-			Name:      "performance_schema_total",
-			Help:      "Mysql instrumentations that could not be loaded or created due to memory constraints",
-		}, []string{"instrumentation"}),
 	}
 }
 
+// Describe implements prometheus.Collector.
 func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
-	for _, m := range e.metrics {
-		m.Describe(ch)
-	}
-
-	e.commands.Describe(ch)
-	e.connectionErrors.Describe(ch)
-	e.innodbRows.Describe(ch)
-	e.performanceSchema.Describe(ch)
+	// We cannot know in advance what metrics the exporter will generate
+	// from MySQL. So we use the poor man's describe method: Run a collect
+	// and send the descriptors of all the collected metrics. The problem
+	// here is that we need to connect to the MySQL DB. If it is currently
+	// unavailable, the descriptors will be incomplete. Since this is a
+	// stand-alone exporter and not used as a library within other code
+	// implementing additional metrics, the worst that can happen is that we
+	// don't detect inconsistent metrics created by this exporter
+	// itself. Also, a change in the monitored MySQL instance may change the
+	// exported metrics during the runtime of the exporter.
+
+	metricCh := make(chan prometheus.Metric)
+	doneCh := make(chan struct{})
+
+	go func() {
+		for m := range metricCh {
+			ch <- m.Desc()
+		}
+		doneCh <- struct{}{}
+	}()
 
-	ch <- e.duration.Desc()
-	ch <- e.totalScrapes.Desc()
-	ch <- e.error.Desc()
+	e.Collect(metricCh)
+	close(metricCh)
+	close(doneCh)
```

Comment:
I think you've misunerstood Fabian here. Line 143 should be `close(doneCh)`, while this line here should still be `<-doneCh`.